applications/GeoMechanicsApplication/custom_strategies/builder_and_solvers/residualbased_block_builder_and_solver_linear_elastic_dynamic.h

```diff
@@ -0,0 +1,735 @@
+                    BaseType::AssembleLHS(mMassMatrix, mass_contribution, equation_ids);
+                }
+                if (damping_contribution.size1() != 0) {
+                    BaseType::AssembleLHS(mDampingMatrix, damping_contribution, equation_ids);
+                }
+
+                // Assemble the elemental contribution
+                BaseType::AssembleLHS(rA, lhs_contribution, equation_ids);
+            }
+        });
+    }
+
+    void CalculateGlobalMatrices(const ConditionsArrayType& rConditions, TSystemMatrixType& rA, ModelPart& rModelPart)
+    {
+        const auto& r_current_process_info = rModelPart.GetProcessInfo();
+
+        block_for_each(rConditions, [&r_current_process_info, &rA, this](Condition& r_condition) {
+            LocalSystemMatrixType lhs_contribution(0, 0);
+            LocalSystemMatrixType mass_contribution(0, 0);
+            LocalSystemMatrixType damping_contribution(0, 0);
+
+            std::vector<std::size_t> equation_ids;
+
+            if (r_condition.IsActive()) {
+                r_condition.CalculateLeftHandSide(lhs_contribution, r_current_process_info);
+                r_condition.CalculateMassMatrix(mass_contribution, r_current_process_info);
+                r_condition.CalculateDampingMatrix(damping_contribution, r_current_process_info);
+
+                r_condition.EquationIdVector(equation_ids, r_current_process_info);
+
+                if (mass_contribution.size1() != 0) {
+                    BaseType::AssembleLHS(mMassMatrix, mass_contribution, equation_ids);
+                }
+                if (damping_contribution.size1() != 0) {
+                    BaseType::AssembleLHS(mDampingMatrix, damping_contribution, equation_ids);
+                }
+
+                // Assemble the elemental contribution
+                BaseType::AssembleLHS(rA, lhs_contribution, equation_ids);
+            }
+        });
+    }
+
+    void AddDynamicsToLhs(TSystemMatrixType& rA, const ModelPart& rModelPart)
+    {
+        const double delta_time = rModelPart.GetProcessInfo()[DELTA_TIME];
+
+        double*       a_values = rA.value_data().begin();
+        const double* m_values = mMassMatrix.value_data().begin();
+        const double* c_values = mDampingMatrix.value_data().begin();
+
+        // add mass and damping contribution to LHS sparse matrix
+        // mass contribution: 1.0 / (mBeta * delta_time * delta_time) * M
+        // damping contribution: mGamma / (mBeta * delta_time) * C
+        for (std::size_t i = 0; i < rA.size1(); i++) {
+            const std::size_t col_begin = rA.index1_data()[i];
+            const std::size_t col_end   = rA.index1_data()[i + 1];
+
+            for (std::size_t j = col_begin; j < col_end; ++j) {
+                a_values[j] += (1.0 / (mBeta * delta_time * delta_time)) * m_values[j];
+                a_values[j] += (mGamma / (mBeta * delta_time)) * c_values[j];
+            }
+        }
+    }
```

Comment:
furthemore AddDynamicsToLHS is a protected method which i cannot acces from here

Reply:
> the function as i implented loops over the sparse matrix, therefore, there are less operations compared to the function in the scheme.
> 
> what do you mean by "re-use the ublas functionality"?

I would expect ublas to have efficient functionality for sparse matrices, such that we could put the following instead of the loops:
```
rA += 1.0/(mBeta * delta_time * delta_time) * mMassMatrix
rA += (mGamma / (mBeta * delta_time)) * mDampingMatrix
```
However, I'm not sure if that's implemented in the efficient way you'd like (and if there are other e.g. size-related issues here), but I think for readability's sake it's worth it to try and see if it doesn't impact performance.


> furthemore AddDynamicsToLHS is a protected method which i cannot acces from here
Indeed, however we might be able to override one of the scheme functions in a new scheme, where we expose this functionality. Since it's generally the scheme that's doing this, I would prefer to keep it there (although we indeed might need adjustments).

Reply:
from what i read on the internet is that the ublas functionality for matrix multiplication is not super efficient. My code is similar as in the base class, which I assume is done for a reason